Detectors/ZDC/simulation/src/Digitizer.cxx

```diff
@@ -551,27 +606,70 @@ void Digitizer::assignTriggerBits(uint32_t ibc, std::vector<BCData>& bcData)
       break;
     }
     const auto& otherBC = bcData[ibc_peek];
+    auto diffBC = otherBC.ir.differenceInBC(currBC.ir);
+    if (diffBC < -1) {
+      continue;
+    }
+    if (diffBC > 3) {
+      break;
+    }
+    if (otherBC.ext_triggers && diffBC >= 0) {
+      for (int im = 0; im < NModules; im++) {
+        currBC.moduleTriggers[im] |= 0x1 << diffBC;
+      }
+    }
     if (otherBC.triggers) {
-      auto diffBC = otherBC.ir.differenceInBC(currBC.ir);
-      if (diffBC < -1 || diffBC > 3) {
-        continue;
+      // Assign trigger bits in payload
+      for (int im = 0; im < NModules; im++) {
+        uint32_t tmask = (0xf << (im * NChPerModule)) & mTriggerMask;
+        if (otherBC.triggers & tmask) {
+          currBC.moduleTriggers[im] |= 0x1 << (5 + diffBC);
+        }
       }
-      diffBC++;
-      if (otherBC.ext_triggers) {
-        for (int im = 0; im < NModules; im++) {
-          currBC.moduleTriggers[im] |= 0x1 << diffBC;
+    }
+  }
+  // Printout before cleanup
+  //currBC.print(mTriggerMask);
+}
+
+void Digitizer::Finalize(std::vector<BCData>& bcData, std::vector<o2::zdc::OrbitData>& pData)
+{
+  // Default is to mask trigger bits, we can leave them for debugging
+  for (Int_t ipd = 0; ipd < pData.size(); ipd++) {
+    for (int id = 0; id < NChannels; id++) {
+      pData[ipd].scaler[id] = 0;
+    }
+  }
+  for (int im = 0; im < NModules; im++) {
+    for (int ic = 0; ic < NChPerModule; ic++) {
+      uint32_t mask = 0x1 << (im * NChPerModule + ic);
+      auto id = mModuleConfig->modules[im].channelID[ic];
+      for (uint32_t ibc = 0; ibc < bcData.size(); ibc++) {
+        auto& currBC = bcData[ibc];
+        if ((currBC.triggers & mask) && (mReadoutMask & mask)) {
+          for (Int_t ipd = 0; ipd < pData.size(); ipd++) {
+            if (pData[ipd].ir.orbit == currBC.ir.orbit) {
+              pData[ipd].scaler[id]++;
+            }
+          }
         }
       }
-      triggers[diffBC] = otherBC.triggers;
-    }
-    // Assign trigger bits in payload
-    for (int im = 0; im < NModules; im++) {
-      uint32_t tmask = (0xf << (im * NChPerModule)) & mTriggerMask;
-      for (int it = 0; it < 5; it++) {
-        if (triggers[it] & tmask) {
-          currBC.moduleTriggers[im] |= 0x1 << (5 + it);
+    }
+  }
+  if (mMaskTriggerBits) {
+    for (uint32_t ibc = 0; ibc < bcData.size(); ibc++) {
+      auto& currBC = bcData[ibc];
+      for (int im = 0; im < NModules; im++) {
+        // Cleanup hits if module has no trigger
+        uint32_t mmask = 0xf << im;
+        if ((currBC.triggers & mmask) == 0) {
+          currBC.triggers&(~mmask);
```

Comment:
this has no effect and macOS CI fails on this. What should be here?

Reply:
1) Done
2) I see. However one is for digits from simulation and the other is for digits from raw. Should I merge the two into one file? Shouldn't the output file name be different for sim->digit and raw->digit?
3) It was a bug.. It's not a common condition, probably it didn't show up in the events that I tested.

Reply:
For the DigitWriter: yes, the names should be different, but they (actually, the default, which is "zdcdigits.root" for digitization and  "o2digit_zdc.root" for decoding) can be passed as an argument. So, the https://github.com/AliceO2Group/AliceO2/blob/d99f406de1020f095ebfa69e7c30d67bf8f02179/Detectors/ZDC/workflow/src/ZDCDigitWriterDPLSpec.cxx#L28-L41 should look like 

````
template <typename T>
using BranchDefinition = MakeRootTreeWriterSpec::BranchDefinition<T>;
DataProcessorSpec getZDCDigitWriterDPLSpec(bool mctruth, bool simVersion)
{
  using InputSpec = framework::InputSpec;
  using MakeRootTreeWriterSpec = framework::MakeRootTreeWriterSpec;
  std::string fnameDef = simVersion ? "zdcdigits.root" : o2digit_zdc.root";
  std::string writerName = simVersion ? "ZDCDigitWriterSim" : "ZDCDigitWriterSimDec";

  return MakeRootTreeWriterSpec(
 writerName,  fnameDef,  "o2sim", 1, 
 BranchDefinition<std::vector<o2::zdc::BCData>>{InputSpec{"digitBCinput", "ZDC", "DIGITSBC"}, "ZDCDigitBC"},
 BranchDefinition<std::vector<o2::zdc::ChannelData>>{InputSpec{"digitChinput", "ZDC", "DIGITSCH"}, "ZDCDigitCh"},
 BranchDefinition<std::vector<o2::zdc::OrbitData>>{InputSpec{"digitPDinput", "ZDC", "DIGITSPD"}, "ZDCDigitOrbit"},
 BranchDefinition<o2::dataformats::MCTruthContainer<o2::zdc::MCLabel>>{InputSpec{"labelinput", "ZDC", "DIGITSLBL"},  "ZDCDigitLabels", mctruth ? 1 : 0})(); 
}
````

Then, call it on https://github.com/AliceO2Group/AliceO2/blob/d99f406de1020f095ebfa69e7c30d67bf8f02179/Steer/DigitizerWorkflow/src/SimpleDigitizerWorkflow.cxx#L526 as 
````
specs.emplace_back(o2::zdc::getZDCDigitWriterDPLSpec(mctruth, true));
````
and  here https://github.com/AliceO2Group/AliceO2/blob/d99f406de1020f095ebfa69e7c30d67bf8f02179/Detectors/ZDC/workflow/src/o2-zdc-raw2digits.cxx#L59
as 
````
specs.emplace_back(o2::zdc::getZDCDigitWriterDPLSpec(false, false));
````

Reply:
Merged functionality into ZDCDigitWriterDPLSpec